### DISCUSSION TITLE
Service에서 다른 Service를 의존하기 vs Repository를 의존하기

### DISCUSSION THREAD

**Discussion by boorownie** [next-step/atdd-subway-map#489] [General]
아마 다들 아시다시피 정답은 없는 문제이기도 하며 
어떤 부분을 중요하게 여기느냐 혹은 어떤 아키텍처의 영향을 많이 받았는가에 따라서 
구현이 달라질 것 으로 보입니다.

미션을 진행하시면서 어떤 기준으로 미션을 구현해주셨는지 편하게 의견을 나눠보면 좋겠습니다!

**Comment by giyeon95** [comment]
개인적으로는 다른 Service를 참조하는 방식은 도메인 지식이 풍부하다면, 이슈가 없을 수도 있겠지만 `순환 참조에 대한 잠재젹 이슈`와, 각 서비스레이어에서 `비즈니스 로직이 의존하고`있을 수도 있지 않을까? 싶은 의견을 가지고 있어요, 비즈니스가 커지면 헷갈리기도 하고.. `Repository` 를 의존하는게 개인적으로는 파악하기 쉽지 않을까 싶은 의견이 있습니다.

다른분들 의견도 궁금하네요🤔

(번외로 중복 코드를 제거하기 위해 Service 레이어를 두가지로 분리하는 시도도 몇번 해보았는데 이것도 의견이 많이 갈리는 것 같습니다..ㅎㅎ)
```
ex) xxxController -> xxxApiService -> xxxService -> xxxRepository
```

**Comment by zbqmgldjfh** [comment]
저도 서비스가 커질수록 Service 계층에서는 Repository만 끌어다 사용하는 것 이 좋다고 생각은 합니다. 
다만, 순환 참조가 생기지 않겠다고 생각드는 부분에서는 Service를 써도 될거같다 생각합니다.

이 주제는 답이 없는 부분인것 같습니다 ㅎㅎ, 판단 기준은 서비스의 환경이 정하는것 같습니다 ㅎㅎ

다음은 백기선님이 남기셨던 답글인데, 확인해보시면 도움이 될것 입니다 ㅎㅎ!!
https://www.inflearn.com/questions/21703

**Comment by realrains** [comment]
당연히 서비스가 다른 서비스를 주입 받을 수 있다고 생각합니다. 서비스에서 의존하고자 하는 역할이 단순한 Repository 로서의 기능이냐, 비즈니스 로직을 포함한 기능이냐에 따라 어떤 대상을 주입할지 적절히 선택해야 하는 것 같아요. 다만, AService 에서 직접적으로 필요없는 기능을 모두 포함한 BService 를 통채로 주입받는 식의 상황은 유지/보수나 테스트 용이성 면에서 지양해야 한다고 생각합니다.

개인적으로는 정말 단순한 CRUD 작업만을 포함한 비즈니스로직을 담은 레이어라면 XXXService 라는 이름의 컴포넌트로 두어도 괜찮다고 생각하지만. 도메인 서비스에 대한 비즈니스 요구사항이 늘어나거나 복잡해 질 수록 서비스 하나가 단순히 도메인들을 오케스트레이션하는 역할을 넘어서서 여러가지 동작을 포함한 슈퍼 클래스가 되기 쉽고, 서비스 내 각각의 메서드에서는 필요없는 의존성들이 하나의 서비스에서 모이기 쉬워 (예를들어 LineService 에 구현된 모든 메서드가 LineRepository 와 StationRepository (or StationService ) 를 동시에 참조하지는 않음) Service 자체 또는 일부를 기능 단위로 컴포넌트(혹은 도메인 서비스)로 분리한 뒤 이를 주입 받아 단방향 의존성 그래프를 이루도록 만드는 것이 좋다고 생각합니다!

**Comment by MangKyu** [comment]
이미 많은 분들이 얘기해주셨지만 결국 정해진 것은 없고, 상황 by 상황이긴 할 것 같습니다! 하지만 많은 분들이 Repository 주입을 옹호하시는 거 같아서 원활한 토론을 위해 Service를 주입받아야 하는 입장에서 적어보겠습니다ㅎㅎ

서비스 계층을 주입받아야 하는 이유는 일반적으로 트랜잭션 경계에 해당되기 때문입니다. "토비의 스프링"에 이와 관련된 내용이 있는데 제가 읽으면서 정리했던 책의 내용들을 공유드리겠습니다!
> 트랜잭션을 중구난방으로 적용하는 것은 좋지 않다. 대신 특정 계층의 경계를 트랜잭션 경계와 일치시키는 것이 좋은데, 일반적으로 비지니스 로직을 담고 있는 서비스 계층의 메소드와 결합시키는 것이 좋다. 왜냐하면 데이터 저장 계층으로부터 읽어온 데이터를 사용하고 변경하는 등의 작업을 하는 곳이 서비스 계층이기 때문이다. 서비스 계층을 트랜잭션의 시작과 종료 경계로 정했다면, 테스트와 같은 특별한 이유가 아니고는 다른 계층이나 모듈에서 DAO에 직접 접근하는 것은 차단해야 한다. 트랜잭션은 보통 서비스 계층의 메소드 조합을 통해 만들어지기 때문에 DAO가 제공하는 주요 기능은 서비스 계층에 위임 메소드를 만들어둘 필요가 있다. 그리고 가능하면 다른 모듈의 DAO에 접근할 때는 서비스 계층을 거치도록 하는 것이 바람직하다.

그 다음은 단순 CRUD만 하는 기능들은 시간이 지나면서 그 이상의 기능들을 갖기 쉬워지기 때문입니다. 이떄의 문제는 단순 CRUD 이상의 기능들이 서비스 계층 내지는 엔티티에 구현되지 않아 로직들이 퍼지기 쉬워지고, 중복이 발생하기 쉽습니다. 예를 들어 처음에는 단순 Read 작업이라 컨트롤러에서 Repository를 주입받아 처리하던 것에 추가적인 처리 작업이 필요한 경우입니다. 서비스 계층을 두고 서비스 계층에서 처리하면 적합한 내용은 Repository를 직접 주입받는 곳(여기서는 컨트롤러)에 작성하게 될 가능성이 높고, 이는 코드를 찾기 어렵게 만들며 중복을 유발하기 쉽습니다. 테스트는 당연히 어려워집니다ㅎㅎ 이와 관련된 내용은 직접 경험 + "스프링부트 실전 활용 마스터"에서 참고했습니다!

앞서 다른 분이 언급하신 순환 참조 이슈는 생성자 주입을 사용했다면 애플리케이션 실행 시점에 즉시 파악이 가능해 문제 없을 것 같습니다! 참고로 스프링 부트 2.6 부터는 필드 주입도 실행 시점에 파악 가능하므로 점점 대응이 쉬워질 것 같습니다! 인수테스트만 돌려도 바로 에러가 날테니까요!
또한 앞서 설명주신대로  비대한 서비스 주입보다는 서비스 클래스를 분리하거나 인터페이스를 분리해 주입받는 것게 동의합니다! 관련해서 재밌는 [링크](https://johngrib.github.io/wiki/article/hierarchical-controller-package-structure/)도 하나 공유해드리겠습니다:) 이 링크 내용에 공감이 갔다면 헥사고날 아키텍처를 봐보셔도 좋을 것 같습니당ㅎㅎ

그렇다면 항상 서비스 계층을 만들어 이를 주입받아야 하는가? 는 또 아니라고 생각합니다! 현재의 아키텍처 구성이나 환경 등의 고려도 필요하겠지만 그런 부분들을 다 적으면 너무 복잡해지므로 문제를 심플하게 가져가면...  단순 CRUD라면 Repository 주입을 허용하되, 기능 추가나 서비스 계층에 담길 로직이 필요해진다면 그 순간을 리팩토링 시점으로 이해하고 빠르게 변경하는게 중요하다고 생각합니다ㅎㅎ 이를 "지름길을 사용한다"라고 표현하기도 하는데, 지름길을 사용했다면 미래의 부채를 줄이기 위해 변경이 필요한 즉시 돌아가는 것 입니다! 물론 이게 쉽지 않고 손이 매우 안가는건 또 문제같습니다ㅎㅎ
그리고 결국 가장 중요한건 Repository든 Service든 이것을 주입받을 때 고려했던 부분과 납득갈만한 이유가 아닐까 싶네요:)
